File: Unity3D/Extra/CSSParse.cs
/*
 * Copyright 2014, Gregg Tavares.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Gregg Tavares. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

using UnityEngine;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace CSSParse {

public class Style {
    static public Color ParseCSSColor(string s) {
        // I have no idea what all the color formats are. The ones here are
        // name:         red,              green,            purple
        // #RGB:         #F00,             #0F0,             #F0F
        // #RRGGBB       #FF0000,          #00FF00,          #FF00FF
        // rgb(r,g,b)    rgb(255,0,0)      rgb(0,255,0),     rgb(255,0,255)
        // rgba(r,g,b,a) rgba(255,0,0,1.0) rgba(0,255,0,1.0) rgba(255,0,255,1.0)

        Color color = new Color();
        Match m;
        m = Style.m_hexrrggbbRE.Match(s);
        if (m.Success) {
            color.r = int.Parse(m.Groups[1].Value, System.Globalization.NumberStyles.HexNumber) / 255.0f;
            color.g = int.Parse(m.Groups[2].Value, System.Globalization.NumberStyles.HexNumber) / 255.0f;
            color.b = int.Parse(m.Groups[3].Value, System.Globalization.NumberStyles.HexNumber) / 255.0f;
            color.a = 1.0f;
            return color;
        }

        m = Style.m_hexrgbRE.Match(s);
        if (m.Success) {
            color.r = int.Parse(m.Groups[1].Value, System.Globalization.NumberStyles.HexNumber) / 15.0f;
            color.g = int.Parse(m.Groups[2].Value, System.Globalization.NumberStyles.HexNumber) / 15.0f;
            color.b = int.Parse(m.Groups[3].Value, System.Globalization.NumberStyles.HexNumber) / 15.0f;
            color.a = 1.0f;
            return color;
        }

        m = Style.m_rgbRE.Match(s);
        if (m.Success) {
            color.r = float.Parse(m.Groups[1].Value) / 255.0f;
            color.g = float.Parse(m.Groups[2].Value) / 255.0f;
            color.b = float.Parse(m.Groups[3].Value) / 255.0f;
            color.a = 1.0f;
            return color;
        }

        m = Style.m_rgbaRE.Match(s);
        if (m.Success) {
            color.r = float.Parse(m.Groups[1].Value) / 255.0f;
            color.g = float.Parse(m.Groups[2].Value) / 255.0f;
            color.b = float.Parse(m.Groups[3].Value) / 255.0f;
            color.b = float.Parse(m.Groups[4].Value);
            return color;
        }

        m = Style.m_nameRE.Match(s);
        if (m.Success) {
            Style.InitColors();
            string name = m.Groups[1].Value;
            string lcName = name.ToLowerInvariant();
            if (!m_colors.TryGetValue(lcName, out color)) {
                Debug.LogError("unknown color: " + name);
            }
            return color;
        }

        Debug.LogError("unsupported color format: " + s);
        return color;
    }

    private static Regex m_hexrrggbbRE = new Regex(@"\s*#([0-9a-fA-F][0-9a-fA-F])([0-9a-fA-F][0-9a-fA-F])([0-9a-fA-F][0-9a-fA-F])\s*");
    private static Regex m_hexrgbRE = new Regex(@"\s*#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])\s*");
    private static Regex m_rgbRE = new Regex(@"\s*rgb\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)\s*");
    private static Regex m_rgbaRE = new Regex(@"\s*rgba\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*,\s*(.+)\)\s*");
    private static Regex m_nameRE = new Regex(@"\s*(\w+)\s*");

    static private Dictionary<string, Color> m_colors;

    private static void InitColors() {
        if (m_colors == null) {
            m_colors = new Dictionary<string, Color>();
            string[] parts = m_colorData.Split(',');
            for (int ii = 0; ii < parts.Length; ii += 2) {
                string name = parts[ii + 0].Trim();
                string colorString = parts[ii + 1].Trim();
                Color color = new Color();
                color.r = int.Parse(colorString.Substring(2, 2), System.Globalization.NumberStyles.HexNumber) / 255.0f;
                color.g = int.Parse(colorString.Substring(4, 2), System.Globalization.NumberStyles.HexNumber) / 255.0f;
                color.b = int.Parse(colorString.Substring(6, 2), System.Globalization.NumberStyles.HexNumber) / 255.0f;
                color.a = int.Parse(colorString.Substring(0, 2), System.Globalization.NumberStyles.HexNumber) / 255.0f;

                m_colors[name] = color;
            }
        }
    }

    // In .NET 2.0 there's no easy way to have static data?
    private static string m_colorData = @"aliceblue,fff0f8ff,
antiquewhite,fffaebd7,
aqua,ff00ffff,
aquamarine,ff7fffd4,
azure,fff0ffff,
beige,fff5f5dc,
bisque,ffffe4c4,
black,ff000000,
blanchedalmond,ffffebcd,
blue,ff0000ff,
blueviolet,ff8a2be2,
brown,ffa52a2a,
burlywood,ffdeb887,
cadetblue,ff5f9ea0,
chartreuse,ff7fff00,
chocolate,ffd2691e,
coral,ffff7f50,
cornflowerblue,ff6495ed,
cornsilk,fffff8dc,
crimson,ffdc143c,
cyan,ff00ffff,
darkblue,ff00008b,
darkcyan,ff008b8b,
darkgoldenrod,ffb8860b,
darkgray,ffa9a9a9,
darkgrey,ffa9a9a9,
darkgreen,ff006400,
darkkhaki,ffbdb76b,
darkmagenta,ff8b008b,
darkolivegreen,ff556b2f,
darkorange,ffff8c00,
darkorchid,ff9932cc,
darkred,ff8b0000,
darksalmon,ffe9967a,
darkseagreen,ff8fbc8f,
darkslateblue,ff483d8b,
darkslategray,ff2f4f4f,
darkslategrey,ff2f4f4f,
darkturquoise,ff00ced1,
darkviolet,ff9400d3,
deeppink,ffff1493,
deepskyblue,ff00bfff,
dimgray,ff696969,
dimgrey,ff696969,
dodgerblue,ff1e90ff,
firebrick,ffb22222,
floralwhite,fffffaf0,
forestgreen,ff228b22,
fuchsia,ffff00ff,
gainsboro,ffdcdcdc,
ghostwhite,fff8f8ff,
gold,ffffd700,
goldenrod,ffdaa520,
gray,ff808080,
grey,ff808080,
green,ff008000,
greenyellow,ffadff2f,
honeydew,fff0fff0,
hotpink,ffff69b4,
indianred,ffcd5c5c,
indigo,ff4b0082,
ivory,fffffff0,
khaki,fff0e68c,
lavender,ffe6e6fa,
lavenderblush,fffff0f5,
lawngreen,ff7cfc00,
lemonchiffon,fffffacd,
lightblue,ffadd8e6,
lightcoral,fff08080,
lightcyan,ffe0ffff,
lightgoldenrodyellow,fffafad2,
lightgray,ffd3d3d3,
lightgrey,ffd3d3d3,
lightgreen,ff90ee90,
lightpink,ffffb6c1,
lightsalmon,ffffa07a,
lightseagreen,ff20b2aa,
lightskyblue,ff87cefa,
lightslateblue,ff8470ff,
lightslategray,ff778899,
lightslategrey,ff778899,
lightsteelblue,ffb0c4de,
lightyellow,ffffffe0,
lime,ff00ff00,
limegreen,ff32cd32,
linen,fffaf0e6,
magenta,ffff00ff,
maroon,ff800000,
mediumaquamarine,ff66cdaa,
mediumblue,ff0000cd,
mediumorchid,ffba55d3,
mediumpurple,ff9370db,
mediumseagreen,ff3cb371,
mediumslateblue,ff7b68ee,
mediumspringgreen,ff00fa9a,
mediumturquoise,ff48d1cc,
mediumvioletred,ffc71585,
midnightblue,ff191970,
mintcream,fff5fffa,
mistyrose,ffffe4e1,
moccasin,ffffe4b5,
navajowhite,ffffdead,
navy,ff000080,
oldlace,fffdf5e6,
olive,ff808000,
olivedrab,ff6b8e23,
orange,ffffa500,
orangered,ffff4500,
orchid,ffda70d6,
palegoldenrod,ffeee8aa,
palegreen,ff98fb98,
paleturquoise,ffafeeee,
palevioletred,ffdb7093,
papayawhip,ffffefd5,
peachpuff,ffffdab9,
peru,ffcd853f,
pink,ffffc0cb,
plum,ffdda0dd,
powderblue,ffb0e0e6,
purple,ff800080,
red,ffff0000,
rosybrown,ffbc8f8f,
royalblue,ff4169e1,
saddlebrown,ff8b4513,
salmon,fffa8072,
sandybrown,fff4a460,
seagreen,ff2e8b57,
seashell,fffff5ee,
sienna,ffa0522d,
silver,ffc0c0c0,
skyblue,ff87ceeb,
slateblue,ff6a5acd,
slategray,ff708090,
slategrey,ff708090,
snow,fffffafa,
springgreen,ff00ff7f,
steelblue,ff4682b4,
tan,ffd2b48c,
teal,ff008080,
thistle,ffd8bfd8,
tomato,ffff6347,
transparent,00000000,
turquoise,ff40e0d0,
violet,ffee82ee,
violetred,ffd02090,
wheat,fff5deb3,
white,ffffffff,
whitesmoke,fff5f5f5,
yellow,ffffff00,
yellowgreen,ff9acd32";
}

/*
        string[] res = {
            "#1af",
            " #1af ",
            "#12abef",
            " #12abef ",
            "rgb(30,100,255)",
            " rgb( 30 , 100 , 255 ) ",
            "rgba(30,100,255,0.5)",
            " rgba( 30 , 100 , 255 , 0.5 ) ",
            "green",
            " purple ",
        };
        for (int ii = 0; ii < res.Length; ++ii) {
            Color x = Style.ParseCSSColor(res[ii]);
            print("" + ii + ": " + x.r + ", " + x.g + ", " + x.b + ", " + x.a);
        }
*/

}  // namespace CSSParse

File: public/hft/0.0.0/scripts/misc/cssparse.js
/*
 * Copyright 2014, Gregg Tavares.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Gregg Tavares. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

define(function() {

  var s_colorData = {
    aliceblue: 0xfff0f8ff,
    antiquewhite: 0xfffaebd7,
    aqua: 0xff00ffff,
    aquamarine: 0xff7fffd4,
    azure: 0xfff0ffff,
    beige: 0xfff5f5dc,
    bisque: 0xffffe4c4,
    black: 0xff000000,
    blanchedalmond: 0xffffebcd,
    blue: 0xff0000ff,
    blueviolet: 0xff8a2be2,
    brown: 0xffa52a2a,
    burlywood: 0xffdeb887,
    cadetblue: 0xff5f9ea0,
    chartreuse: 0xff7fff00,
    chocolate: 0xffd2691e,
    coral: 0xffff7f50,
    cornflowerblue: 0xff6495ed,
    cornsilk: 0xfffff8dc,
    crimson: 0xffdc143c,
    cyan: 0xff00ffff,
    darkblue: 0xff00008b,
    darkcyan: 0xff008b8b,
    darkgoldenrod: 0xffb8860b,
    darkgray: 0xffa9a9a9,
    darkgrey: 0xffa9a9a9,
    darkgreen: 0xff006400,
    darkkhaki: 0xffbdb76b,
    darkmagenta: 0xff8b008b,
    darkolivegreen: 0xff556b2f,
    darkorange: 0xffff8c00,
    darkorchid: 0xff9932cc,
    darkred: 0xff8b0000,
    darksalmon: 0xffe9967a,
    darkseagreen: 0xff8fbc8f,
    darkslateblue: 0xff483d8b,
    darkslategray: 0xff2f4f4f,
    darkslategrey: 0xff2f4f4f,
    darkturquoise: 0xff00ced1,
    darkviolet: 0xff9400d3,
    deeppink: 0xffff1493,
    deepskyblue: 0xff00bfff,
    dimgray: 0xff696969,
    dimgrey: 0xff696969,
    dodgerblue: 0xff1e90ff,
    firebrick: 0xffb22222,
    floralwhite: 0xfffffaf0,
    forestgreen: 0xff228b22,
    fuchsia: 0xffff00ff,
    gainsboro: 0xffdcdcdc,
    ghostwhite: 0xfff8f8ff,
    gold: 0xffffd700,
    goldenrod: 0xffdaa520,
    gray: 0xff808080,
    grey: 0xff808080,
    green: 0xff008000,
    greenyellow: 0xffadff2f,
    honeydew: 0xfff0fff0,
    hotpink: 0xffff69b4,
    indianred: 0xffcd5c5c,
    indigo: 0xff4b0082,
    ivory: 0xfffffff0,
    khaki: 0xfff0e68c,
    lavender: 0xffe6e6fa,
    lavenderblush: 0xfffff0f5,
    lawngreen: 0xff7cfc00,
    lemonchiffon: 0xfffffacd,
    lightblue: 0xffadd8e6,
    lightcoral: 0xfff08080,
    lightcyan: 0xffe0ffff,
    lightgoldenrodyellow: 0xfffafad2,
    lightgray: 0xffd3d3d3,
    lightgrey: 0xffd3d3d3,
    lightgreen: 0xff90ee90,
    lightpink: 0xffffb6c1,
    lightsalmon: 0xffffa07a,
    lightseagreen: 0xff20b2aa,
    lightskyblue: 0xff87cefa,
    lightslateblue: 0xff8470ff,
    lightslategray: 0xff778899,
    lightslategrey: 0xff778899,
    lightsteelblue: 0xffb0c4de,
    lightyellow: 0xffffffe0,
    lime: 0xff00ff00,
    limegreen: 0xff32cd32,
    linen: 0xfffaf0e6,
    magenta: 0xffff00ff,
    maroon: 0xff800000,
    mediumaquamarine: 0xff66cdaa,
    mediumblue: 0xff0000cd,
    mediumorchid: 0xffba55d3,
    mediumpurple: 0xff9370db,
    mediumseagreen: 0xff3cb371,
    mediumslateblue: 0xff7b68ee,
    mediumspringgreen: 0xff00fa9a,
    mediumturquoise: 0xff48d1cc,
    mediumvioletred: 0xffc71585,
    midnightblue: 0xff191970,
    mintcream: 0xfff5fffa,
    mistyrose: 0xffffe4e1,
    moccasin: 0xffffe4b5,
    navajowhite: 0xffffdead,
    navy: 0xff000080,
    oldlace: 0xfffdf5e6,
    olive: 0xff808000,
    olivedrab: 0xff6b8e23,
    orange: 0xffffa500,
    orangered: 0xffff4500,
    orchid: 0xffda70d6,
    palegoldenrod: 0xffeee8aa,
    palegreen: 0xff98fb98,
    paleturquoise: 0xffafeeee,
    palevioletred: 0xffdb7093,
    papayawhip: 0xffffefd5,
    peachpuff: 0xffffdab9,
    peru: 0xffcd853f,
    pink: 0xffffc0cb,
    plum: 0xffdda0dd,
    powderblue: 0xffb0e0e6,
    purple: 0xff800080,
    red: 0xffff0000,
    rosybrown: 0xffbc8f8f,
    royalblue: 0xff4169e1,
    saddlebrown: 0xff8b4513,
    salmon: 0xfffa8072,
    sandybrown: 0xfff4a460,
    seagreen: 0xff2e8b57,
    seashell: 0xfffff5ee,
    sienna: 0xffa0522d,
    silver: 0xffc0c0c0,
    skyblue: 0xff87ceeb,
    slateblue: 0xff6a5acd,
    slategray: 0xff708090,
    slategrey: 0xff708090,
    snow: 0xfffffafa,
    springgreen: 0xff00ff7f,
    steelblue: 0xff4682b4,
    tan: 0xffd2b48c,
    teal: 0xff008080,
    thistle: 0xffd8bfd8,
    tomato: 0xffff6347,
    transparent: 0x00000000,
    turquoise: 0xff40e0d0,
    violet: 0xffee82ee,
    violetred: 0xffd02090,
    wheat: 0xfff5deb3,
    white: 0xffffffff,
    whitesmoke: 0xfff5f5f5,
    yellow: 0xffffff00,
    yellowgreen: 0xff9acd32,
  };

  var s_hexrrggbbRE = /\s*#([0-9a-fA-F][0-9a-fA-F])([0-9a-fA-F][0-9a-fA-F])([0-9a-fA-F][0-9a-fA-F])\s*/;
  var s_hexrgbRE = /\s*#([0-9a-fA-F])([0-9a-fA-F])([0-9a-fA-F])\s*/;
  var s_rgbRE = /\s*rgb\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)\s*/;
  var s_rgbaRE = /\s*rgba\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*,\s*(.+)\)\s*/;
  var s_nameRE = /\s*(\w+)\s*/;

  // given a CSS color string, returns an array of 4 integers [r, g, b, a] in the range 0 to 255.
  // if opt_0to1 is passed in the values will be 0 to 1.
  var parseCSSColor = function(s, opt_0to1) {
    // I have no idea what all the color formats are. The ones here are
    // name:         red,              green,            purple
    // #RGB:         #F00,             #0F0,             #F0F
    // #RRGGBB       #FF0000,          #00FF00,          #FF00FF
    // rgb(r,g,b)    rgb(255,0,0)      rgb(0,255,0),     rgb(255,0,255)
    // rgba(r,g,b,a) rgba(255,0,0,1.0) rgba(0,255,0,1.0) rgba(255,0,255,1.0)

    var m;
    var c;
    m = s_hexrrggbbRE.exec(s);
    if (m) {
      c = [
        parseInt(m[1], 16),
        parseInt(m[2], 16),
        parseInt(m[3], 16),
        255,
      ];
    } else if (m = s_hexrgbRE.exec(s)) {
      var r = parseInt(m[1], 16);
      var g = parseInt(m[2], 16);
      var b = parseInt(m[3], 16);
      c = [
        r * 16 + r,
        g * 16 + g,
        b * 16 + b,
        255,
      ];
    } else if (m = s_rgbRE.exec(s)) {
      c = [
        parseInt(m[1]),
        parseInt(m[2]),
        parseInt(m[3]),
        255,
      ];
    } else if (m = s_rgbaRE.exec(s)) {
      c = [
        parseInt(m[1]),
        parseInt(m[2]),
        parseInt(m[3]),
        Math.floor(parseFloat(m[4]) * 255),
      ];
    } else if(m = s_nameRE.exec(s)) {
      var name = m[1].toLowerCase();
      var color = s_colorData[name];
      if (color !== undefined) {
        c = [
          (color >> 16) & 0xFF,
          (color >>  8) & 0xFF,
          (color >>  0) & 0xFF,
          (color >> 24) & 0xFF,
        ];
      }
    }

    if (!c) {
      console.error("unsupported color format: " + s);
    }

    if (opt_0to1) {
      c[0] = c[0] / 255;
      c[1] = c[1] / 255;
      c[2] = c[2] / 255;
      c[3] = c[3] / 255;
    }

    return c;
  };

  /*
  var res = [
      "#1af",
      " #1af ",
      "#12abef",
      " #12abef ",
      "rgb(30,100,255)",
      " rgb( 30 , 100 , 255 ) ",
      "rgba(30,100,255,0.5)",
      " rgba( 30 , 100 , 255 , 0.5 ) ",
      "green",
      " purple ",
  ];
  console.log("here!");
  for (var ii = 0; ii < res.length; ++ii) {
    var a = parseCSSColor(res[ii]);
    console.log("" + ii + ": " + a);
  }
  */

  return {
    parseCSSColor: parseCSSColor,
  };

});
